mantidproject/mantid
# Enable SystemTests failures to be processed/mass updated

**Description of work.**
This PR enables further processing of SystemTest failures. This is focussed around making it easy look at each failure by first linking each test failure to its generated mismatch workspace. With this information it is possible to create an output json file
When the validation method is WorkspaceToNexus, a full dictionary entry for the failed test is created, composing of:
```
FailedTest {
'TestName': Name of the test which failed,
'MismatchFile': Path to the mismatch file generated by the test,
'ReferenceFile': Path to reference file,
'ReferenceContentLinkFiles': Path to the original content link of the reference file
}
```

As this information provides a complete view of the file(s) required by the test, it can be used to replace the original test data using a newly generated python script addTestData.py:

`python path_to_script/Testing/SystemTests/addTestData.py --testresults FAILED-systemtests.json --replacereferences   `

This will do the following:
-	Create a new content and content link file based on the mismatch file
-	Replace the old content link file with the new one
-	Tell the user to upload the content (data) file to the test server
It can operate on multiple test failure, so can be used to mass update reference files (as long as the validation method was WorkspaceToNexus). 

Furthermore, this json file can be read in workbench to load each mismatch and reference file to compare/analyse any differences between them. 

This script can also be used to add test data, e.g if we had our new test files in a directory "testfiles":
```
/testfiles
    test1.nxs
    test2.nxs
    test3.nxs
```
 and we wanted to create our content link (hash files) in a directory ref we could do
`python path_to_script\addTestData.py --inputfiles testfiles\*.nxs --targetdirectory ref`

Which would create a content (data) file for each of the input nexus files and then create the content link (hash) file in a directory called ref
```
/ref
    test1.nxs.md5
    test2.nxs.md5
    test3.nxs.md5
```
Which you would then move into the appropriate mantid test data folder and upload the content (data) files to the Mantid data server. 



**To test:**
- Open to any feedback on this e.g. in terms of usability of the generated script
- You'll probably have to create a failing system test to test it: I was doing the following
- Open the VesuvioFittingTest system test
- Swap around the validate methods for VesuvioFittingWithQuadraticBackgroundTest and VesuvioFittingWithKFreeTest
- Build and run the system test
- Two tests will fail and a json file should be generated in the SystemTest save dir e.g. mantid_build/Testing/SystemTests/scripts:
```
{
  "FailedTests": [
    {
      "TestName": "VesuvioFittingTest.VesuvioFittingWithKFreeTest",
      "MismatchFiles": [
        "WithQuadraticBackgroundTest-mismatch.nxs"
      ],
      "ReferenceFiles": [
        "VesuvioFittingWithQuadraticBackgroundTest.nxs"
      ],
      "ReferenceContentLinkFiles": [
        "VesuvioFittingWithQuadraticBackgroundTest.nxs.md5"
      ]
    },
    {
      "TestName": "VesuvioFittingTest.VesuvioFittingWithQuadraticBackgroundTest",
      "MismatchFiles": [
        "VesuvioFittingWithKFreeTest-mismatch.nxs"
      ],
      "ReferenceFiles": [
        "VesuvioFittingWithKFreeTest.nxs"
      ],
      "ReferenceContentLinkFiles": [
        " VesuvioFittingWithKFreeTest.nxs.md5"
      ]
    }
  ]
}
```
- If you open a terminal here and do:
`python path_to_script/Testing/SystemTests/addTestData.py --testresults FAILED-systemtests.json --replacereferences   `
It should update the VesuvioFittingWithKFreeTest.nxs.md5 and VesuvioFittingWithQuadraticBackgroundTest.nxs.md5 files in the mantid source directory with a new hash (which is the hash of the mismatch workspace).

Fixes #30273 . 



*This does not require release notes* because it is an internal change.




---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
